schemes.go

```diff
@@ -921,6 +925,84 @@ func (scheme *SchemeManager) parseContents(conf *Configuration) error {
 		conf.Issuers[issuer.Identifier()] = issuer
 		return scheme.parseCredentialsFolder(conf, issuer, filepath.Join(dir, "Issues"))
 	})
+
+	if err != nil {
+		return err
+	}
+
+	// validate that there are no circular dependencies
+	for _, credType := range conf.CredentialTypes {
+		if credType.SchemeManagerID == scheme.ID {
+			if err := credType.validateDependencies(conf, []CredentialTypeIdentifier{}, credType.Identifier()); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (ct CredentialType) validateDependencies(conf *Configuration, validatedDeps DependencyChain, toBeChecked CredentialTypeIdentifier) error {
+	if len(validatedDeps) >= maxDepComplexity {
+		return errors.New("dependency tree too complex: " + validatedDeps.String())
+	}
+	for _, outer := range conf.CredentialTypes[ct.Identifier()].Dependencies {
+		// at least one middle needs to be valid
+		atLeastOneInnerConSatisfied := false
+		for _, middle := range outer {
+			innerConSatisfied := true
+
+			for _, item := range middle {
+				if conf.CredentialTypes[item].SchemeManagerID != ct.SchemeManagerID {
+					return errors.Errorf("credential type %s in scheme %s has dependency outside the scheme: %s",
+						ct.Identifier().String(), ct.SchemeManagerID, conf.CredentialTypes[item].Identifier().String())
+				}
+
+				// all items need to be valid for middle to be valid
+				if toBeChecked == item {
+					innerConSatisfied = false
+					break
+				}
+
+				if conf.CredentialTypes[item].Dependencies != nil {
+					if e := conf.CredentialTypes[item].validateDependencies(conf, append(validatedDeps, item), toBeChecked); e != nil {
+						if e.Error() == "No valid dependency branch could be built. There might be a circular dependency." {
+							innerConSatisfied = false
+						} else {
+							return e
+						}
+					}
+				}
+			}
+
+			if innerConSatisfied {
+				atLeastOneInnerConSatisfied = true
```

Comment:
Here too you can `break` now, right?

Reply:
introducing a break here would not further have the rest of the branches not be validated for being too complex